RecoLocalTracker/SiPixelClusterizer/plugins/alpaka/SiPixelRawToClusterKernel.h

```diff
@@ -0,0 +1,199 @@
+  const uint32_t ladderMask = 0xFF;
+  const uint32_t moduleMask = 0x3FF;
+  const uint32_t panelMask = 0x3;
+  const uint32_t diskMask = 0xF;
+  const uint32_t bladeMask = 0x3F;
+
+  const uint32_t LINK_bits = 6;
+  const uint32_t ROC_bits = 5;
+  const uint32_t DCOL_bits = 5;
+  const uint32_t PXID_bits = 8;
+  const uint32_t ADC_bits = 8;
+
+  // special for layer 1
+  const uint32_t LINK_bits_l1 = 6;
+  const uint32_t ROC_bits_l1 = 5;
+  const uint32_t COL_bits_l1 = 6;
+  const uint32_t ROW_bits_l1 = 7;
+  const uint32_t OMIT_ERR_bits = 1;
+
+  const uint32_t maxROCIndex = 8;
+  const uint32_t numRowsInRoc = 80;
+  const uint32_t numColsInRoc = 52;
+
+  const uint32_t MAX_WORD = 2000;
+
+  const uint32_t ADC_shift = 0;
+  const uint32_t PXID_shift = ADC_shift + ADC_bits;
+  const uint32_t DCOL_shift = PXID_shift + PXID_bits;
+  const uint32_t ROC_shift = DCOL_shift + DCOL_bits;
+  const uint32_t LINK_shift = ROC_shift + ROC_bits_l1;
+  // special for layer 1 ROC
+  const uint32_t ROW_shift = ADC_shift + ADC_bits;
+  const uint32_t COL_shift = ROW_shift + ROW_bits_l1;
+  const uint32_t OMIT_ERR_shift = 20;
+
+  const uint32_t LINK_mask = ~(~uint32_t(0) << LINK_bits_l1);
+  const uint32_t ROC_mask = ~(~uint32_t(0) << ROC_bits_l1);
+  const uint32_t COL_mask = ~(~uint32_t(0) << COL_bits_l1);
+  const uint32_t ROW_mask = ~(~uint32_t(0) << ROW_bits_l1);
+  const uint32_t DCOL_mask = ~(~uint32_t(0) << DCOL_bits);
+  const uint32_t PXID_mask = ~(~uint32_t(0) << PXID_bits);
+  const uint32_t ADC_mask = ~(~uint32_t(0) << ADC_bits);
+  const uint32_t ERROR_mask = ~(~uint32_t(0) << ROC_bits_l1);
+  const uint32_t OMIT_ERR_mask = ~(~uint32_t(0) << OMIT_ERR_bits);
+
+  struct DetIdGPU {
+    uint32_t RawId;
+    uint32_t rocInDet;
+    uint32_t moduleId;
+  };
+
+  struct Pixel {
+    uint32_t row;
+    uint32_t col;
+  };
+
+  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE constexpr pixelchannelidentifierimpl::Packing packing() {
+    return PixelChannelIdentifier::thePacking;
+  }
+
+  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE constexpr uint32_t pack(uint32_t row,
+                                                              uint32_t col,
+                                                              uint32_t adc,
+                                                              uint32_t flag = 0) {
+    constexpr pixelchannelidentifierimpl::Packing thePacking = packing();
+    adc = std::min(adc, uint32_t(thePacking.max_adc));
+
+    return (row << thePacking.row_shift) | (col << thePacking.column_shift) | (adc << thePacking.adc_shift);
+  }
+
+  constexpr uint32_t pixelToChannel(int row, int col) {
+    constexpr pixelchannelidentifierimpl::Packing thePacking = packing();
+    return (row << thePacking.column_width) | col;
+  }
+
+}  // namespace pixelDetails
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+  namespace pixelDetails {
+
+    template <typename TrackerTraits>
+    class SiPixelRawToClusterKernel {
+    public:
+      class WordFedAppender {
```

Comment:
`WordFedAppender` seems to be independent of `TrackerTraits`, so it could be moved outside of the `SiPixelRawToClusterKernel` (that should lead to a little bit less code to be generated).